tidy3d/components/geometry/base.py

```diff
@@ -2011,6 +2161,344 @@ def _surface_area(self, bounds: Bound) -> float:
+
+    def _volume(self, bounds: Bound) -> float:
+        """Returns object's volume within given bounds."""
+        # NOTE (Lucas): Bounds are overestimated.
+        vertices = np.dot(self.inverse, self._vertices_from_bounds(bounds))[:3]
+        inverse_bounds = (tuple(vertices.min(axis=1)), tuple(vertices.max(axis=1)))
+        return abs(np.linalg.det(self.transform)) * self.geometry.volume(inverse_bounds)
+
+    def _surface_area(self, bounds: Bound) -> float:
+        """Returns object's surface area within given bounds."""
+        log.warning("Surface area of transformed elements cannot be calculated.")
+        return 0.0
+
+    @staticmethod
+    def translation(x: float, y: float, z: float) -> MatrixReal4x4:
+        """Return a translation matrix.
+
+        Parameters
+        ----------
+        x : float
+            Translation along x.
+        y : float
+            Translation along y.
+        z : float
+            Translation along z.
+
+        Returns
+        -------
+        numpy.ndarray
+            Transform matrix with shape (4, 4).
+        """
+        return np.array(
+            [
+                (1.0, 0.0, 0.0, x),
+                (0.0, 1.0, 0.0, y),
+                (0.0, 0.0, 1.0, z),
+                (0.0, 0.0, 0.0, 1.0),
+            ],
+            dtype=float,
+        )
+
+    @staticmethod
+    def scaling(x: float, y: float = None, z: float = None) -> MatrixReal4x4:
+        """Return a scaling matrix.
+
+        Parameters
+        ----------
+        x : float
+            Scaling factor along x.
+        y : float = None
+            Scaling factor along y. If ``None``, use the value for x.
+        z : float = None
+            Scaling factor along z. If ``None``, use the value for x.
+
+        Returns
+        -------
+        numpy.ndarray
+            Transform matrix with shape (4, 4).
+        """
+        if y is None:
+            y = x
+
+        if z is None:
+            z = x
+
+        if np.isclose((x, y, z), 0.0).any():
+            raise Tidy3dError("Scaling factors cannot be zero in any dimensions.")
+        return np.array(
+            [
+                (x, 0.0, 0.0, 0.0),
+                (0.0, y, 0.0, 0.0),
+                (0.0, 0.0, z, 0.0),
+                (0.0, 0.0, 0.0, 1.0),
+            ],
+            dtype=float,
+        )
+
+    @staticmethod
+    def rotation(angle: float, axis: Union[Axis, Coordinate]) -> MatrixReal4x4:
+        """Return a rotation matrix.
+
+        Parameters
+        ----------
+        angle : float
+            Rotation angle (in radians).
+        axis : Union[int, Tuple[float, float, float]]
+            Axis of rotation: 0, 1, or 2 for x, y, and z, respectively, or a 3D vector.
+
+        Returns
+        -------
+        numpy.ndarray
+            Transform matrix with shape (4, 4).
+        """
+        sin = np.sin(angle)
+        cos = np.cos(angle)
+        if axis == 0:
```

Comment:
this might give an error if axis is a numpy array I think. Maybe we should handle this case by converting axis to np.ndarray and then `tolist()` or something like this? 

<img width="568" alt="image" src="https://github.com/flexcompute/tidy3d/assets/92756888/c4bf9035-7783-4ee7-843d-76122fd9ef2c">

Reply:
Performing the parameter validation before the comparison seems to work too, and avoids the type juggling. What do you think?